lambdajs/src/SharedInterpreter.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE Arrows                     #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+module SharedInterpreter where
+
+import           Control.Arrow
+import           Control.Arrow.Utils (map)
+import           Prelude             hiding (break, error, lookup, map, read)
+import qualified Prelude
+import           Syntax
+
+class Arrow c => AbstractValue v c | c -> v where
+    -- values
+    numVal :: c Double v
+    boolVal :: c Bool v
+    stringVal :: c String v
+    undefVal :: c () v
+    nullVal :: c () v
+    lambdaVal :: c ([Ident], Expr) v
+    objectVal :: c [(Ident, v)] v
+    getField :: c Expr v -> c (v, Expr) v
+    updateField :: c Expr v -> c (v, Expr, v) v
+    deleteField :: c Expr v -> c (v, Expr) v
+    -- operator/delta function
+    evalOp :: c (Op, [v]) v
+    -- environment ops
+    lookup :: c Ident v
+    apply :: c Expr v -> c (v, [v]) v
+    -- store ops
+    set :: c (v, v) ()
+    new :: c v v
+    get :: c v v
+    -- control flow
+    if_ :: c Expr v -> c Expr v -> c (v, Expr, Expr) v
+    while_ :: c Expr v -> c Expr v -> c (Expr, Expr) v
```

Comment:
While loops can be implemented in shared code:
https://github.com/svenkeidel/sturdy/blob/master/while/src/SharedSemantics.hs#L87-L88

One operation less for which we have to worry about soundness.

Reply:
True, this is how I implemented `while` loops before, but this would always cause infinite loops. Maybe this could be implemented together with the fixpoint arrow transformer.